rcl/include/rcl/introspection.h

```diff
@@ -0,0 +1,162 @@
+#include "rcl/time.h"
+#include "rmw/rmw.h"
+#include "rcl/client.h"
+
+#define RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX "/_service_event"
+
+typedef struct rcl_service_introspection_utils_s {
+  rcl_clock_t * clock;
+  rcl_publisher_t * publisher;
+  rosidl_message_type_support_t * request_type_support;
+  rosidl_message_type_support_t * response_type_support;
+  char * service_name;
+  char * service_type_name;
+  char * service_event_topic_name;
+  bool _enabled;
+  bool _content_enabled;
+} rcl_service_introspection_utils_t;
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_service_introspection_utils_t
+rcl_get_zero_initialized_introspection_utils();
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_introspection_init(
+  rcl_service_introspection_utils_t * introspection_utils,
+  const rosidl_service_type_support_t * service_type_support,
+  const char * service_name,
+  const rcl_node_t * node,
+  rcl_allocator_t * allocator);
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_introspection_fini(
+  rcl_service_introspection_utils_t * introspection_utils,
+  rcl_allocator_t * allocator,
+  rcl_node_t *  node);
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_introspection_send_message(
+  const rcl_service_introspection_utils_t * introspection_utils,
+  const uint8_t event_type,
+  const void * ros_response_request,
+  const int64_t sequence_number,
+  const uint8_t uuid[16], // uuid is uint8_t but the guid is int8_t
+  const rcl_allocator_t * allocator);
+
+
+
+
+
+/*  Enables service introspection by reconstructing the introspection clock and publisher
+ *  
+ *  Does nothing and returns RCL_RET_OK if already enabled
+ */
+RCL_LOCAL
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_introspection_enable(
+    rcl_service_introspection_utils_t * introspection_utils,
+    const rcl_node_t * node,
+    rcl_allocator_t * allocator);
+
+/*  Disabled service introspection by fini-ing and freeing the introspection clock and publisher
+ *  
+ *  Does nothing and returns RCL_RET_OK if already disabled
+ *
+ *
+ *
+ */
+
+RCL_LOCAL
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_introspection_disable(
+    rcl_service_introspection_utils_t * introspection_utils,
+    rcl_node_t * node,
+    const rcl_allocator_t * allocator);
+
+
+
+
+/*
+ * Enables/disables service introspection for client/service
+ * These functions are thin wrappers around rcl_service_introspection_{enable, disable}
+ *
+ *
+ *
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_introspection_configure_service_events(
+    rcl_service_t * service,
+    rcl_node_t * node,
+    bool enable);
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_introspection_configure_client_events(rcl_client_t * client,
+    rcl_node_t * node,
+    bool enable);
+
+
+/*
+ *
+ */
+RCL_PUBLIC
+void
+rcl_service_introspection_configure_client_content(rcl_client_t * client, bool enable);
+
+RCL_PUBLIC
+void
+rcl_service_introspection_configure_service_content(rcl_service_t * service, bool enable);
```

Comment:
Same for this two, but I also don't understand what they are for....

Reply:
These two functions are for configuring whether or not to include the Request/Response event message in the introspected output. For example someone might only want to introspect the timing of Request/Response events and so does not want to incur the added overhead of publishing out (potentially large) service event messages.

Reply:
> These two functions are for configuring whether or not to include the Request/Response event message in the introspected output. For example someone might only want to introspect the timing of Request/Response events and so does not want to incur the added overhead of publishing out (potentially large) service event messages.

Ah, I see.
I think it makes more sense to have only one `rcl_service_events_options_t`, which could be a member of [rcl_service_options_t](https://github.com/ros2/rcl/blob/4eccc3cbe65f5b58981f22efaf612a65731cb2f0/rcl/include/rcl/service.h#L43-L50), instead of having many functions to enable disable the events and to enable disable sending the messages contents.

We should also improve naming, something like "message payload" would be more clear than "service content".

Reply:
+1 for `message_payload`, or maybe the more verbose `request_response_payload`